server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java

```diff
@@ -165,6 +170,31 @@ public void testInternalSendExceptionForksToGenericIfHandlerDoesNotFork() {
 
             assertEquals("simulated exception in sendRequest", getSendRequestException(future, IOException.class).getMessage());
         }
+        assertWarnings(
+            "[transport.enable_stack_protection] setting was deprecated in Elasticsearch and will be removed in a future release."
+        );
+    }
+
+    public void testInternalSendExceptionWithNonForkingResponseHandlerCanCompleteFutureWaiterFromGenericThread() throws Exception {
+        try (var nodeA = new TestNode("node-A")) {
+            final var future = new PlainActionFuture<TransportResponse.Empty>();
+            final var latch = new CountDownLatch(1);
+            nodeA.transportService.getThreadPool().generic().execute(() -> {
+                assertEquals("simulated exception in sendRequest", getSendRequestException(future, IOException.class).getMessage());
+                latch.countDown();
+            });
```

Comment:
Not sure what this bit of the test is doing - is it needed?

Reply:
The original reason that leads to this change is that PlainActionFuture#assertCompleteAllowed` fails in `PlainActionFuture#onFailure` when the waiter is with a `generic` thread. So I wanted to have this test to simulate that situation, i.e. waiting for future with a generic thread. 

I do recognize that asserting the completing thread is currentThread instead of `generic` already proves the change is effective. But I added the above intending to preserve some context on how this issue is identified. If you think that's unnecessary, I can definitely rewrite the test as you suggested. Please let me know.

Reply:
Sorry I just noticed the latch is indeed useless. Removed in ffaefbdc4987fe0ce25dfba6e35706746586266e

Reply:
(this is an overall response right, not just related to the few lines at the top of this thread?)

Yeah I worry that this is a very indirect way to test that property, and it'd still be a problem if we called `sendRequest` on a (different) generic thread from the one that's waiting on the future. I'd rather we just focussed on the actual threading behaviour here in the context of the `TransportService` tests.

Reply:
> Yeah I worry that this is a very indirect way to test that property

I agree. It's in addition to the same thread assertion. 

You are right that assertion can still be tripped if we send the request with the generic thread pool. This is already another problem that can happen in stress test (with a slightly different code path). 

I pushed e90005d0e221abecbf5efe7233b5449c324c1cbb to rewrite the test as you suggested.